core/src/main/java/io/questdb/griffin/engine/functions/json/JsonExtractFunction.java

```diff
@@ -0,0 +1,441 @@
+                    return Numbers.LONG_NULL;
+                }
+            case SimdJsonType.NUMBER: {
+                return extractLongFromJsonNumber(res);
+            }
+            default:
+                return Numbers.LONG_NULL;
+        }
+    }
+
+    @Override
+    public final int getType() {
+        return columnType;
+    }
+
+    @Override
+    public @Nullable DirectUtf8Sink getVarcharA(Record rec) {
+        return extractVarchar(json.getVarcharA(rec), stateA);
+    }
+
+    @Override
+    public @Nullable DirectUtf8Sink getVarcharB(Record rec) {
+        assert stateB != null;
+        return extractVarchar(json.getVarcharB(rec), stateB);
+    }
+
+    @Override
+    public int getVarcharSize(Record rec) {
+        final Utf8Sequence utf8seq = getVarcharA(rec);
+        if (utf8seq == null) {
+            return TableUtils.NULL_LEN;
+        }
+        return utf8seq.size();
+    }
+
+    @Override
+    public void init(SymbolTableSource symbolTableSource, SqlExecutionContext executionContext) throws SqlException {
+        json.init(symbolTableSource, executionContext);
+        path.init(symbolTableSource, executionContext);
+        pointer = Misc.free(pointer);
+        pointer = JsonExtractSupportingState.varcharConstantToJsonPointer(path);
+    }
+
+    private long extractLongFromJsonNumber(long res) {
+        switch (stateA.simdJsonResult.getNumberType()) {
+            case SimdJsonNumberType.SIGNED_INTEGER:
+                return res;
+            case SimdJsonNumberType.FLOATING_POINT_NUMBER: {
+                final double d = Double.longBitsToDouble(res);
+                if (d >= Long.MIN_VALUE && d <= Long.MAX_VALUE) {
+                    return (long) d;
+                }
+                break;
+            }
+        }
+        return Numbers.LONG_NULL;
+    }
+
+    private @Nullable DirectUtf8Sink extractVarchar(Utf8Sequence json, JsonExtractSupportingState state) {
+        if (json != null && pointer != null) {
+            assert state.destUtf8Sink != null;
+            state.destUtf8Sink.clear();
+            state.parser.queryPointerUtf8(
+                    state.initPaddedJson(json),
+                    pointer,
+                    state.simdJsonResult,
+                    state.destUtf8Sink,
+                    maxSize
+            );
+
+            if (state.simdJsonResult.hasValue()) {
+                return state.destUtf8Sink;
+            }
+            state.throwIfInError(jsonPosition);
+        }
+        return null;
+    }
+
+    private long queryPointerValue(Utf8Sequence jsonInput) {
+        assert stateA.destUtf8Sink != null;
+        stateA.destUtf8Sink.clear();
+        final long res = stateA.parser.queryPointerValue(
+                stateA.initPaddedJson(jsonInput),
+                pointer,
+                stateA.simdJsonResult,
+                stateA.destUtf8Sink,
+                maxSize
+        );
+
+        if (!stateA.simdJsonResult.hasValue()) {
```

Comment:
This isn't _quite_ accurate.
The code here should use logic similar to `throwIfInError` instead.

The caller should still check the return value to determine if it should return a default value if a minor error has occurred like an `INDEX_OUT_OF_BOUNDS` or a `NO_SUCH_FIELD` error.

As alternative, the function `queryPointerValue` fn could also take a `long defaulValue` so it can then:
  - Reset the `SimdJsonResult` to `getError() == .SUCCESS`, `getType() == .NUMBER` and `getNumericType() == .SIGNED_INTEGER`.
  - Return it a default value.

This second approach would alleviate the need for the caller to perform yet fiddly success/failure checking.